File: Baekjoon/9663N-Queen.cpp
#include<iostream>
#include<cmath>
using namespace std;

int arr[15];
int n;
int cnt = 0;

bool promising(int row) {
	for (int i = 0; i < row; i++) {
		/*직선이나 대각선에 있는지 확인*/
		if (arr[i] == arr[row] || abs(i - row) == abs(arr[i] - arr[row])) {
			return false;
		}
	}
	return true;
}

void NQueen(int row) {
	if (row == n) {
		cnt++;
		return;
	}
	for (int col = 0; col < n; col++) {
		arr[row] = col;
		if (promising(row)) {
			NQueen(row + 1);
		}
	}
}

int main() {
	cin >> n;
	NQueen(0);
	cout << cnt;
	return 0;
}

File: Baekjoon/9663n-queen.py
def promising(r):
    for i in range(r):
        if r - i == abs(arr[r] - arr[i]):
            return False
    return True

def solve(r):
    if r == n:
        global result
        result += 1
        return

    for i in range(n):
        if visit[i]:
            continue
        arr[r] = i
        if promising(r):
            visit[i] = True
            solve(r + 1)
            visit[i] = False



n = int(input())
arr = [0] * n
visit = [False] * n
result = 0
solve(0)
print(result)
